### DISCUSSION TITLE
Release Notes: v1.14.0

### DISCUSSION THREAD

**Discussion by PavelBlend** [PavelBlend/blender-xray#426] [General]
Список нововведений в версии 1.14.0.

## Улучшения:

 - Добавлен оператор `Resize Bones` для изменения размера костей. Оператор находится в окне `3D Viewport`, в N панеле.
![Screenshot_1](https://user-images.githubusercontent.com/7983249/131881112-3426a055-b63e-4b8f-8e1d-14bd4476ec34.jpg) 
Оператор имеет параметры:
![04](https://user-images.githubusercontent.com/7983249/131880991-a9a92448-2a4b-499d-9d2e-226461b36344.jpg)

    - `Mode` - режим. Всего два режима:
        - `Constant` - размер костей одинаков для всех костей.
        - `Adaptive` - размер костей расчитывается индивидуально для каждой кости.
    - `Bone Size` - размер кости для режима `Constant`. В режиме `Adaptive`, этот параметр не оказывает никакого влияния.
    - `Custom Shapes` - если этот параметр включён, то костям будет назначен `Custom Shape` в виде пустышки с типом `Sphere`.

 - В N панель окна `3D Viewport` в панель `Batch Tools` добавлен оператор `Place Selected Objects` для изменения позиции у выделенных объектов. При массовом импорте (когда импортируется больше одного файла) все модели и скелеты импортируются в начало координат. С помощью этого оператора можно расположить модели в разные точки. Аддон изменяет позицию только root-объектов (объекты, у которых включен параметр `Object` в окне `Properties`, панели `X-Ray Engine: Object`). Объекты располагаются в алфавитном порядке. У оператора есть параметры:
    - `Rows` - количество рядов, на которых нужно расположить модели.
    - `Horizontal Offset` - расстояние между моделями по оси X.
    - `Vertical Offset` - расстояние между моделями по оси Z.
 
После batch-импорта объекты находятся в одной точке:
![05](https://user-images.githubusercontent.com/7983249/131881807-d8217c24-c368-424c-ac6f-f29a95dc5a1e.jpg)

После вызова оператора `Place Selected Objects`:
![06](https://user-images.githubusercontent.com/7983249/131881777-2937b845-0894-4129-8c54-b01576c99992.jpg)

 - При экспорте скелетных моделей, аддон объединяет меш-объекты в один объект. До этого, аддон выдавал предупреждение `Skeletal object has more than one mesh` и останавливал экспорт. Сейчас аддон при экспорте будет создавать временный объект (который удалится после экспорта), который является результатом объединения скелетных мешей. И экспорт будет выполнен корректно (если не обединять скелетные меши в один объект, то `SDK` будет выдавать ошибку). Если аддон делает объединение мешей, то будет выведено предупреждение `mesh-objects have been merged`.

 - Теперь в списке игровых материалов аддон отображает описания, которые прописаны в `gamemtl.xr` файле.
![Screenshot_2](https://user-images.githubusercontent.com/7983249/131882182-36ab8cd0-29d0-4562-b2ff-454b9e0fb345.jpg)

## Изменения:

 - Удалён оператор `Custom Shape for Bones` при импорте `*.object` файлов. Теперь для изменения размеров костей нужно использовать оператор `Resize Bones`.

 - Добавлено предупреждение при импорте `*.skls`, когда скелет имеет недопустимый режим вращения. Анимация будет импортирована, но нужно будет вручную изменить режим вращения костей. В `*.skls` для костей используется режим вращения `ZXY Euler`. Если кости в блендере имеют другой режим вращения, то импортированная `*.skls` анимация будет не правильной. Чтобы она была правильной, нужно всем костям установить режим вращения `ZXY Euler`. Делать реимпорт `*.skls` анимаций не нужно. Достаточно после импорта у костей изменить режим вращения.

 - Удалён параметр `Selection Only` у оператора экспорта в `*.ogf` формат. Теперь аддон экспортирует в `*.ogf` только выделенные объекты. Экспортёр всё так же поддерживает экспорт только одного root-объекта (объект, у которого включён параметр `Object`).

 - Изменено отображение параметров в панеле `Object`, раздел `Details`. Теперь имена параметров не заслоняются полями ввода.

## Исправления:

 - Исправлена ошибка при импорте `*.skls`, когда в `*.skls` файле отсутствовали некоторые кости. В предыдущей версии аддона в такой ситуации появлялась ошибка.

 - Добавлено предупреждение при экспорте анимации в `*.anm`, когда объект не имеет анимационных ключей для осей X, Y, Z. До этого аддон выдавал нечитаемую ошибку. Теперь, если объект не имеет анимационных ключей хотя бы для одной оси вращения или позиции, аддон выдаст читаемое предупреждение и анимация не будет экспортирована.

**Comment by mortany** [comment]
Класс, это теперь можно экспортировать несколько мешей с одним скелетом как одно целое?

**Reply by PavelBlend** [reply]
@mortany да, для удобства добавил. Но и без этой возможности не особо много проблем было. Я всегда держал все скелетные меши в одном объекте. Или перед экспортом объединял.
